style: Allow sharing styles across elements with presentational hints.

emilio opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17063)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/sharing/mod.rs, components/style/rule_tree/mod.rs, components/style/context.rs, components/style/matching.rs, components/style/traversal.rs and 2 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio commented:
r? @bholley

emilio commented:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=97726f83bbda0c08ed821d2bd22a9e0dd71412cb&selectedJob=102756194

(Note that the jpeg test failure is because the try run is on top of my patches for bug 1357583, which introduce that who knows why)

bholley commented on components/style/context.rs:
```diff
@@ -154,26 +153,28 @@ impl<'a> SharedStyleContext<'a> {
-    /// a good reason to.
+    /// The element being processed. Currently we use an OpaqueNode since we
+    /// only use this for identity checks, but we could use SendElement if there
+    /// were a good reason to.
```
Can you please stop adding gratuitous whitespace reformats to your patches? It makes them much harder to review, which means they take longer, which means it takes longer for me to find time to review them, which ends up as review delay. I don't think either of us want that. :-)

bholley commented on components/style/context.rs:
```diff
@@ -154,26 +153,28 @@ impl<'a> SharedStyleContext<'a> {
-    /// revalidation selectors.
-    pub revalidation_match_results: Option<BitVec>,
+    /// Lazy cache of the stuff we use to share style, in case we tried and
+    /// failed.
```
Change this comment to:

"When we evaluate whether two elements may share style, we have to compute various values and compare them. We cache this information lazily here so that it's only computed once per element.

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -32,6 +34,88 @@ pub enum StyleSharingBehavior {
+/// `StyleSharingCandidate` (indeed we could try to merge those, though
+/// conceptually we may still want them to be separate).
+#[derive(Debug)]
+pub struct CachedStyleSharingData {
```
Let's call this StyleSharingCriteria.

emilio replied:
I don't love "criteria", I don't think it says much, this is just a bunch of data we cache with the purpose of style sharing. But fine, I guess.

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -109,6 +218,103 @@ pub struct StyleSharingCandidateCache<E: TElement> {
+    /// because it depends on the `style_sharing_candidate_cache` having only
+    /// live nodes in it, and we have no way to guarantee that at the type
+    /// system level yet.
+    pub unsafe fn share_style_if_possible(
```
This comment and the associated unsafety are bogus now that we we stored this all in ScopedTLS. Feel free to fix.

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -32,6 +34,88 @@ pub enum StyleSharingBehavior {
+        if self.class_list.is_none() {
+            let mut ret = SmallVec::new();
+            element.each_class(|c| ret.push(c.clone()));
+            self.class_list = Some(ret);
```
Can we sort this list please? Otherwise we may falsely reject candidates based on different ordering.

emilio replied:
I think that's something we may want to land/measure separately.

bholley replied:
How come? Sorting a SmallVec of a few atoms is going to be super-cheap...

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -45,10 +129,35 @@ pub struct StyleSharingCandidate<E: TElement> {
-    /// The cached result of matching this entry against the revalidation selectors.
-    revalidation_match_results: Option<BitVec>,
+    /// The cached data we store to avoid recomputing too much stuff.
+    cache: CachedStyleSharingData,
```
Along with the rename of the struct suggested elsewhere, we should call this (and other names) |criteria| or |sharing_criteria|.

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -109,6 +218,103 @@ pub struct StyleSharingCandidateCache<E: TElement> {
+/// A struct we use as a simple cache of stuff we use to test against different
+/// candidates.
+#[derive(Debug)]
+pub struct ElementSharingChecker<E: TElement> {
```
I'm not really sold on the need for this class. It really seems to me like it's just a bunch of extra boilerplate and indirection, and that we'd be better off by just passing around the element and cache directly.

emilio replied:
I think it's nice because it allows you to keep the sharing flags stuff in a well-known comment, but I need to rewrite all this anyway, so we'll see :)

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -175,10 +373,16 @@ impl<E: TElement> StyleSharingCandidateCache<E> {
+        // If we know the element had no pres hints from the relations, we can
+        // just do as-if we had computed it, regardless of whether we actually
+        // did.
+        if !relations.intersects(AFFECTED_BY_PRESENTATIONAL_HINTS) {
```
This seems like a hack. How about eliminating AFFECTED_BY_PRESENTATIONAL_HINTS altogether, and instead having push_applicable_declarations store the synthesis result directly in the StyleSharingCriteria hanging off the CurrentElementInfo?

bholley commented on components/style/sharing/mod.rs:
```diff
@@ -228,11 +431,6 @@ impl<E: TElement> StyleSharingCandidateCache<E> {
             return StyleSharingResult::CannotShare
         }
 
-        if element.get_id().is_some() {
```
Wait what? Where did this come from? This seems like something out of Boris' patch.

emilio replied:
This should've landed when Boris landed the `element.get_id() != candidate.get_id()` changes not long ago.

bholley reviewed (commented):
So, I really appreciate all the awesome work here, but I'm going to finally bite the bullet and complain about the megapatches because they're driving me crazy. :-)

In general I would ask that you you do the following things, at least for large patches that you want me to review:
(1) Don't include gratuitous reformatting or whitespace change along with other non-trivial changes.
(2) Split up your commits in a way that makes larger change more incrementally.
(3) Post the patches themselves to bugzilla, not just a link to a github PR.

I know this takes more time, but the current approach has the following disadvantages:
(1) I can't review your patches without blocking out a large portion of time, because there's no way for me to track what I've reviewed and what I haven't. This leads to review delays.
(2) The patches take me much longer to review than they otherwise would, which costs me time. The review is also lower quality, because there are too many disjoint things going on, and so it's harder for me to engage with the patch at a precise level and hard not to just hand-wave it.
(3) If I want to r- anything, I have to review the entire patch again.

I've been hesitant to complain about this because you're doing such great work, and I know you almost always get it right. But reviewing your code is costing me way more time than it could, and time is something I'm very short on right now. :-)

emilio commented:
Re. the mega-patches, I'd say that it's fair, but given how long does it take a patch to land, I usually prefer to just stick most related changes together. And re. everything in one commit, that was just that it took less work, and I'm also relatively short on time.

But fair, landing times should be better now, so I'll split this :)

bholley commented:
> but given how long does it take a patch to land, I usually prefer to just stick most related changes together.

You can still push an arbitrary number of changesets, even from single bugs, in the same PR.

> And re. everything in one commit, that was just that it took less work, and I'm also relatively
> short on time.

I know, which is why I've been keeping quiet about it, but this review took me more than 40 minutes to do (ignoring the time I spent complaining about the megapatches), and I still don't feel like I did a very good job. I could have reviewed it in a quarter of the time if it was split up.

bholley commented:
I could do ValidationInfo or ValidationData instead of criteria.

emilio committed:
style: Allow StyleSource to be compared.

emilio committed:
style: Allow ApplicableDeclarationBlocks to be compared.

We'll need this to cache pres hints, which generate these.

emilio committed:
style: rewrap some comments to 80 columns for consistency with the rest of the code.

emilio committed:
style: Add a CachedStyleSharingData to hold the candidate class list and revalidation results.

emilio committed:
style: Also cache the class list in the CurrentElementInfo.

This patch also removes all notion of style sharing from matching.rs, which is
nice.

emilio committed:
style: Allow sharing style for elements with presentational hints.

emilio committed:
style: Avoid some cfg checks in context.rs

emilio commented:
Rewritten, just read the comments re. `ValidationInfo`. Need to run now, happy to rename it in a fixup commit. Feel free to review assuming that.

bholley reviewed (commented):
r=me with those things fixed. Thanks for splitting things up! Next time feel free to combine the PartialEq impls into the commits that use them, and please submit the patches to bugzilla, because github's tracking of revisions for multiple patches is pretty broken. :-)

bholley commented:
Oh, and the most important review comment seems to have been swallowed by github: Please fix things up to cache the presentational hints synthesized in the stylist and avoid re-synthesizing them for the cache.

@bors-servo delegate+

bors-servo commented:
:v: @emilio can now approve this pull request

emilio commented on components/style/sharing/mod.rs:
```diff
@@ -50,6 +51,23 @@ pub struct CachedStyleSharingData {
+
+    /// Move the cached data to a new instance, and return it.
+    pub fn take(&mut self) -> Self {
+        Self {
```
How can this miss any member, exactly?

emilio committed:
s/CachedStyleSharingData/ValidationData.

I still think CachedStyleSharingData should be the name, but not going to fight
over it.

emilio commented:
I left this on IRC, but just so it's not missed:

```
11:09 <emilio> bholley: I looked into the caching pres hints thing... I don't really like the result, complexity wise. I've been thinking on replacing most of the arguments to push_applicable_declarations with something else that abstracts the getting animation rules, style attribute, etc, so maybe that allows to do it in a clean way
11:10 <emilio> bholley: anyway, I don't really want to complicate code on purpose for something that we a) don't have perf numbers for, and b) is already really rare (most elements don't have any pres hints)
```

@bors-servo r=bholley

bors-servo commented:
:pushpin: Commit 7db2776 has been approved by `bholley`

<!-- @bors-servo r=bholley 7db27763488b5fe9661ae9ec52f497c89db16905 -->

bors-servo commented:
:hourglass: Testing commit 7db27763488b5fe9661ae9ec52f497c89db16905 with merge 38a6a3bff6f6e0e35eb592891d5e70e7cb897b69...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7224), [arm32](http://build.servo.org/builders/arm32/builds/7439), [arm64](http://build.servo.org/builders/arm64/builds/7476), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7568), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4324), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4343), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7547), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/32), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/31), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4252), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4317), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/32), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/33), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2791)
Approved by: bholley
Pushing 38a6a3bff6f6e0e35eb592891d5e70e7cb897b69 to master...
